### PR TITLE
Row filtering

### LINKED ISSUE #2270: Changes to ParquetRecordBatchStream to support row filtering in DataFusion
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The current (non-public) API for skipping records has some shortcomings when trying to implement predicate pushdown from DataFusion. It assumes that we have the entire `VecDequeue<RowSelection>` for an entire row group when construction the `ParquetRecordBatchReader` which makes doing streaming filtering a challenge. 

**Describe the solution you'd like**

I wrote a prototype of row-level filtering in DataFusion that required a few changes to the API here. Two main things:

1. Add a method to `ParquetRecordBatchStream`

```
pub fn next_selection(
        &mut self,
        selection: &mut VecDeque<RowSelection>,
    ) -> Option<ArrowResult<RecordBatch>>
```

which mimics `poll_next` but applies a selection to the next batch.

2. Add a method to `ParquetRecordBatchReader`

```
pub fn next_selection(
        &mut self,
        selection: &mut VecDeque<RowSelection>,
    ) -> Option<ArrowResult<RecordBatch>>
```   
which mimics `next` but applies a selection and, importantly, emits a single batch containing all selected rows in `selection`. The current implementation just emits the next `RowSelection` that selects and stops. 

See https://github.com/coralogix/arrow-datafusion/blob/row-filtering/datafusion/core/src/physical_plan/file_format/parquet.rs for the (very hacky) protoype in DataFusion built on https://github.com/coralogix/arrow-rs/tree/row-filtering.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #2270 

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/arrow_reader.rs [hunk 1]
```diff
@@ -283,42 +284,112 @@ pub struct ParquetRecordBatchReader {
...
+
+                // TODO Why does this cause problems?
+                // if skipped != front.row_count {
+                //     return Some(Err(general_err!(
+                //         "failed to skip rows, expected {}, got {}",
+                //         front.row_count,
+                //         skipped
+                //     )
+                //     .into()));
+                // }
```
Comment (thinkharderdev) [inline parquet/src/arrow/arrow_reader.rs:311]: I'm not sure what's up here. This is the correct assertion to make here but I think there is a bug somewhere in the `skip` logic that skips the right values but gets the accounting wrong. Leaving this check in causes issues and leaving it out seems to produce the correct results (if the selections are created correctly).

**Comment by thinkharderdev** [comment]
@tustvold @alamb @Ted-Jiang

File: parquet/src/arrow/arrow_reader.rs (hunk 1 above)
Comment (alamb) [reply parquet/src/arrow/arrow_reader.rs:311]: @tustvold  can you help out here? I haven't been following along with this PR / work closely enough to provide useful guidance here?

**Comment by tustvold** [comment]
@thinkharderdev how would you like to proceed with this, in particular with regards to the suggestion on the ticket of pushing this into the ParquetRecordBatchReader?

**Comment by thinkharderdev** [comment]
> @thinkharderdev how would you like to proceed with this, in particular with regards to the suggestion on the ticket of pushing predicates into the ParquetRecordBatchReader? I would be happy to stub out an API for this if that would be helpful?

I can take a crack at it and post a draft PR when the API is nailed down

**Comment by thinkharderdev** [comment]
Closing in favor of https://github.com/apache/arrow-rs/pull/2310

**Pull request closed without merging**